modules/grpc/clickhouse/clickhouse-dest-worker.cpp

```diff
@@ -24,25 +24,188 @@
 
 using syslogng::grpc::clickhouse::DestWorker;
 using syslogng::grpc::clickhouse::DestDriver;
 
 DestWorker::DestWorker(GrpcDestWorker *s)
   : syslogng::grpc::DestWorker(s)
 {
+  std::shared_ptr<::grpc::ChannelCredentials> credentials = this->create_credentials();
+  if (!credentials)
+    {
+      msg_error("Error querying ClickHouse credentials",
+                evt_tag_str("url", this->owner.get_url().c_str()),
+                log_pipe_location_tag(&this->super->super.owner->super.super.super));
+      throw std::runtime_error("Error querying ClickHouse credentials");
+    }
+
+  ::grpc::ChannelArguments args = this->create_channel_args();
+
+  this->channel = ::grpc::CreateCustomChannel(this->owner.get_url(), credentials, args);
+  this->stub = ::clickhouse::grpc::ClickHouse::NewStub(this->channel);
+}
+
+bool
+DestWorker::should_initiate_flush()
+{
+  return this->current_batch_bytes >= this->get_owner()->batch_bytes;
 }
 
 LogThreadedResult
 DestWorker::insert(LogMessage *msg)
 {
+  DestDriver *owner_ = this->get_owner();
+  std::streampos last_pos = this->query_data.tellp();
+  size_t row_bytes = 0;
+
+  google::protobuf::Message *message = owner_->schema.format(msg, this->super->super.seq_num);
+  if (!message)
+    goto drop;
+
+  this->batch_size++;
+
+  if (!google::protobuf::util::SerializeDelimitedToOstream(*message, &this->query_data))
+    goto drop;
+
+  row_bytes = this->query_data.tellp() - last_pos;
+  this->current_batch_bytes += row_bytes;
+  log_threaded_dest_driver_insert_msg_length_stats(this->super->super.owner, row_bytes);
+
+  msg_trace("Message added to ClickHouse batch", log_pipe_location_tag(&this->super->super.owner->super.super.super));
+
+  delete message;
+
+  if (!this->client_context.get())
+    {
+      this->client_context = std::make_unique<::grpc::ClientContext>();
+      prepare_context_dynamic(*this->client_context, msg);
+    }
+
+  if (this->should_initiate_flush())
+    return log_threaded_dest_worker_flush(&this->super->super, LTF_FLUSH_NORMAL);
+
+  return LTR_QUEUED;
+
+drop:
+  if (!(owner_->template_options.on_error & ON_ERROR_SILENT))
+    {
+      msg_error("Failed to format message for ClickHouse, dropping message",
+                log_pipe_location_tag(&this->super->super.owner->super.super.super));
+    }
+
+  /* LTR_DROP currently drops the entire batch */
+  return LTR_QUEUED;
+}
+
+void
+DestWorker::prepare_query_info(::clickhouse::grpc::QueryInfo &query_info)
+{
+  DestDriver *owner_ = this->get_owner();
+
+  query_info.set_database(owner_->get_database());
+  query_info.set_user_name(owner_->get_user());
+  query_info.set_password(owner_->get_password());
+  query_info.set_query(owner_->get_query());
+  query_info.set_input_data(this->query_data.str());
+}
+
+static LogThreadedResult
+_map_grpc_status_to_log_threaded_result(const ::grpc::Status &status)
+{
+  // TODO: this is based on OTLP, we should check how the ClickHouse gRPC server behaves
```

Comment:
Fix this todo.

Reply:
We agreed on merging as is, so this can be closed.